### PR TITLE
planner: write statements to read on TiFlash only in strict sql mode

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/43090

Problem Summary:
See https://github.com/pingcap/tidb/issues/43090

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -1761,104 +1763,105 @@ func NewSessionVars(hctx HookContext) *SessionVars {
...
+		WaitSplitRegionTimeout:                 DefWaitSplitRegionTimeout,
+		enableIndexMerge:                       DefTiDBEnableIndexMerge,
+		NoopFuncsMode:                          TiDBOptOnOffWarn(DefTiDBEnableNoopFuncs),
+		replicaRead:                            kv.ReplicaReadLeader,
+		AllowRemoveAutoInc:                     DefTiDBAllowRemoveAutoInc,
+		UsePlanBaselines:                       DefTiDBUsePlanBaselines,
+		EvolvePlanBaselines:                    DefTiDBEvolvePlanBaselines,
+		EnableExtendedStats:                    false,
+		IsolationReadEngines:                   make(map[kv.StoreType]struct{}),
+		TiFlashEngineRemovedDueToStrictSQLMode: false,
```
[the remark this answers is not in the document]
Comment (gengliqi) [reply sessionctx/variable/session.go]: I move it to `SessionVars.StmtCtx`.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**